pkg/handlers/generic/lifecycle/clusterautoscaler/strategy_crs.go

```diff
@@ -0,0 +1,142 @@
+	)
+
+	flags.StringVar(
+		&c.defaultClusterAutoscalerConfigMap,
+		prefix+".default-cluster-autoscaler-configmap-name",
+		"cluster-autoscaler",
+		"name of the ConfigMap used to deploy cluster-autoscaler",
+	)
+}
+
+type crsStrategy struct {
+	config crsConfig
+
+	client ctrlclient.Client
+}
+
+func (s crsStrategy) apply(
+	ctx context.Context,
+	req *runtimehooksv1.AfterControlPlaneInitializedRequest,
+	log logr.Logger,
+) error {
+	defaultCM := &corev1.ConfigMap{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: corev1.SchemeGroupVersion.String(),
+			Kind:       "ConfigMap",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: s.config.defaultsNamespace,
+			Name:      s.config.defaultClusterAutoscalerConfigMap,
+		},
+	}
+
+	err := s.client.Get(
+		ctx,
+		ctrlclient.ObjectKeyFromObject(defaultCM),
+		defaultCM,
+	)
+	if err != nil {
+		return fmt.Errorf("failed to get default cluster-autoscaler ConfigMap: %w", err)
+	}
+
+	log.Info("Ensuring cluster-autoscaler ConfigMap exists for cluster")
+
+	cluster := &req.Cluster
+
+	data := templateData(defaultCM.Data, cluster.Name, cluster.Namespace)
+	cm := &corev1.ConfigMap{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: corev1.SchemeGroupVersion.String(),
+			Kind:       "ConfigMap",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: cluster.Namespace,
+			Name:      defaultCM.Name + "-" + cluster.Name,
+		},
+		Data: data,
+	}
+
+	if err := client.ServerSideApply(ctx, s.client, cm); err != nil {
+		return fmt.Errorf(
+			"failed to apply cluster-autoscaler installation ConfigMap: %w",
+			err,
+		)
+	}
+
+	existingManagementCluster, err := capiutils.ManagementCluster(ctx, s.client)
+	if err != nil {
+		return fmt.Errorf(
+			"failed to get management Cluster: %w",
+			err,
+		)
+	}
+	// The cluster-autoscaler is different from other addons.
+	// It requires all resources to be created in the management cluster,
+	// which means creating the ClusterResourceSet always targeting the management cluster.
+	// In most cases, target the management cluster.
+	// But if existingManagementCluster is nil, i.e. when s.client points to a bootstrap cluster,
+	// target the cluster and assume that will become the management cluster.
+	targetCluster := existingManagementCluster
+	if targetCluster == nil {
+		targetCluster = cluster
+	}
+
+	// In the case when existingManagementCluster is nil, i.e. when s.client points to a bootstrap cluster,
+	// it is possible that the namespace where the cluster will be moved to, and where the cluster-autoscaler resources
+	// will be created, does not exist yet.
+	// In that case, we need to create the namespace in the target cluster.
+	clusterKey := ctrlclient.ObjectKeyFromObject(targetCluster)
+	remoteClient, err := remote.NewClusterClient(ctx, "", s.client, clusterKey)
+	if err != nil {
+		return fmt.Errorf("error creating remote cluster client: %w", err)
+	}
+	if err = utils.EnsureNamespace(ctx, remoteClient, cluster.Namespace); err != nil {
```

Comment:
I think we can just embed the namespace resource in the CRS configmap as we do for other addons.

Reply:
The reason I didn't do that here is because for the other addons this namespace is created on a fresh workload cluster where the namespace wouldn't exist.
In this case, the namespace will be created on the management cluster where this namespace can be shared across multiple clusters or exiting namespaces that have labels and annotations.

Reply:
Good point!